README.md
# remark-preset-gfm

[![Build status](https://img.shields.io/travis/imcuttle/remark-preset-gfm/master.svg?style=flat-square)](https://travis-ci.org/imcuttle/remark-preset-gfm)
[![Test coverage](https://img.shields.io/codecov/c/github/imcuttle/remark-preset-gfm.svg?style=flat-square)](https://codecov.io/github/imcuttle/remark-preset-gfm?branch=master)
[![NPM version](https://img.shields.io/npm/v/remark-preset-gfm.svg?style=flat-square)](https://www.npmjs.com/package/remark-preset-gfm)
[![NPM Downloads](https://img.shields.io/npm/dm/remark-preset-gfm.svg?style=flat-square&maxAge=43200)](https://www.npmjs.com/package/remark-preset-gfm)
[![Prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://prettier.io/)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg?style=flat-square)](https://conventionalcommits.org)

[Remark](https://github.com/remarkjs/remark) preset for gfm (GitHub Favorite Markdown)

## Features

- Github break line
- Gemoji
- Svg head link anchor
- Code highlight

## Installation

```bash
npm install remark-preset-gfm
# or use yarn
yarn add remark-preset-gfm
```

## Usage

```javascript
const remark = require('remark')
const html = require('remark-html')
const gfm = require('remark-preset-gfm')

const markdownHtml = remark()
  .use(gfm)
  .use(html)
  .processSync('markdown text')
  .toString()

const htmlOutput = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
  <link rel="stylesheet" href="//unpkg.com/github-markdown-css">
  <link rel="stylesheet" href="//unpkg.com/highlight.js/styles/github.css">
  <style>
    .markdown-body {
      max-width: 65%;
      margin: 40px auto 30px;
    }
    @media screen and (max-width: 1200px) {
      .markdown-body {
        max-width: 78%;
      }
    }

    @media screen and (max-width: 700px) {
      .markdown-body {
        max-width: 90%;
        margin: 20px auto 10px;
      }
    }
    @media screen and (max-width: 400px) {
      .markdown-body {
        max-width: 98%;
        margin: 20px auto 10px;
      }
    }
  </style>
</head>
<body>
<div class="markdown-body">
  ${markdownHtml}
</div>
</body>
</html>
`
```

## Related

- [github-markdown-css](https://github.com/sindresorhus/github-markdown-css) - The minimal amount of CSS to replicate the GitHub Markdown style
- [Highlight.js github style](https://github.com/highlightjs/highlight.js/blob/master/src/styles/github.css)

## Contributing

- Fork it!
- Create your new branch:  
  `git checkout -b feature-new` or `git checkout -b fix-which-bug`
- Start your magic work now
- Make sure npm test passes
- Commit your changes:  
  `git commit -am 'feat: some description (close #123)'` or `git commit -am 'fix: some description (fix #123)'`
- Push to the branch: `git push`
- Submit a pull request :)

## Authors

This library is written and maintained by imcuttle, <a href="mailto:moyuyc95@gmail.com">moyuyc95@gmail.com</a>.

## License

MIT - [imcuttle](https://github.com/imcuttle) 🐟


CHANGELOG.md
<a name="1.0.1"></a>

## 1.0.1 (2018-10-27)


__tests__/main.test.js
/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */

const remark = require('remark')
const html = require('remark-html')
const fs = require('fs')

const gfm = require('../')
const { fixture } = require('./helper')

function read(name) {
  return fs.readFileSync(fixture(name)).toString()
}

function getChildren(ast) {
  return ast && ast.children
}

describe('remarkPresetGfm', function() {
  it('should autolink-headings', function() {
    expect(
      remark()
        .use(gfm)
        .use(html)
        .use(require('../strip-position'))
        .processSync(read('autolink-headings.md'))
        .toString()
    ).toMatchSnapshot()
  })

  it('should gemoji-to-emoji', function() {
    expect(
      remark()
        .use(gfm)
        .use(require('../strip-position'))
        .processSync(read('gemoji-to-emoji.md'))
        .toString()
    ).toMatchSnapshot()
  })

  it('should frontmatter', function() {
    expect(
      remark()
        .use(gfm)
        .use(require('../strip-position'))
        .use(html)
        .processSync(read('frontmatter.md'))
        .toString()
    ).toMatchSnapshot()
  })

  it('should github-break', function() {
    expect(
      remark()
        .use(gfm)
        .use(require('../strip-position'))
        .use(html)
        .processSync(read('github-break.md'))
        .toString()
    ).toMatchInlineSnapshot(`
"<p>newline<br>
newline<br>
endline</p>
"
`)
  })

  it('should highlight.md', function() {
    expect(
      remark()
        .use(gfm)
        .use(require('../strip-position'))
        .use(html)
        .processSync(read('highlight.md'))
        .toString()
    ).toMatchInlineSnapshot(`
"<pre><code class=\\"hljs language-javascript\\"><span class=\\"hljs-built_in\\">console</span>.log(<span class=\\"hljs-number\\">123</span>)</code></pre>
<pre><code>function a() {}
</code></pre>
"
`)
  })
})


gen-link-content.js
/**
 * @file gen-link-content
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2018/10/27
 *
 */
const unified = require('unified')
const rehypeParse = require('rehype-parse')
const fs = require('fs')

const ast = unified()
  .use(rehypeParse, { fragment: true })
  .use(require('./strip-position'))
  .parse(
    '<svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>'
  ).children

fs.writeFileSync('./link-content.json', JSON.stringify(ast, null, 2))


gen-output.js
/**
 * @file example
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2018/10/27
 *
 */

const fs = require('fs')
const remark = require('remark')
const raw = require('rehype-raw')
const remark2Rehype = require('remark-rehype')
const html = require('rehype-stringify')

const markdownHtml = String(
  remark()
    .use(require('./'))
    .use(remark2Rehype)
    .use(raw)
    .use(html)
    .processSync(fs.readFileSync('./README.md'))
)

const htmlOutput = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
  <title>remark-preset-gfm Example</title>
  <link rel="stylesheet" href="//unpkg.com/github-markdown-css">
  <link rel="stylesheet" href="//unpkg.com/highlight.js/styles/github.css">
  <style>
    .markdown-body {
      max-width: 65%;
      margin: 40px auto 30px;
    }
    @media screen and (max-width: 1200px) {
      .markdown-body {
        max-width: 78%;
      }
    }

    @media screen and (max-width: 700px) {
      .markdown-body {
        max-width: 90%;
        margin: 20px auto 10px;
      }
    }
    @media screen and (max-width: 400px) {
      .markdown-body {
        max-width: 98%;
        margin: 20px auto 10px;
      }
    }
  </style>
</head>
<body>
<div class="markdown-body">
  ${markdownHtml}
</div>
</body>
</html>
`

fs.writeFileSync('./output.html', htmlOutput)
